File: services/analyzer/src/analyzer/store.py
from __future__ import annotations

import asyncio
import json
import sqlite3
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from analyzer.models import AnalysisResultData


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


@dataclass(frozen=True)
class AnalysisRecord:
    id: str
    git_url: str
    ref: str
    status: str  # RUNNING | SUCCEEDED | FAILED
    result: AnalysisResultData | None
    error: str


class SQLiteStore:
    def __init__(self, db_path: Path) -> None:
        self._db_path = db_path

    async def init(self) -> None:
        self._db_path.parent.mkdir(parents=True, exist_ok=True)

        def _init() -> None:
            with sqlite3.connect(self._db_path) as conn:
                conn.execute("PRAGMA journal_mode=WAL;")
                conn.execute(
                    """
                    CREATE TABLE IF NOT EXISTS analyses (
                      id TEXT PRIMARY KEY,
                      git_url TEXT NOT NULL,
                      ref TEXT NOT NULL,
                      status TEXT NOT NULL,
                      created_at TEXT NOT NULL,
                      updated_at TEXT NOT NULL,
                      result_json TEXT,
                      error TEXT
                    )
                    """
                )
                conn.commit()

        await asyncio.to_thread(_init)

    async def create(self, *, id: str, git_url: str, ref: str) -> None:
        now = _now_iso()

        def _create() -> None:
            with sqlite3.connect(self._db_path) as conn:
                conn.execute(
                    """
                    INSERT INTO analyses (id, git_url, ref, status, created_at, updated_at, result_json, error)
                    VALUES (?, ?, ?, ?, ?, ?, NULL, '')
                    """,
                    (id, git_url, ref, "RUNNING", now, now),
                )
                conn.commit()

        await asyncio.to_thread(_create)

    async def set_succeeded(self, *, id: str, result: AnalysisResultData) -> None:
        now = _now_iso()
        payload = json.dumps(result.to_json_dict(), separators=(",", ":"), ensure_ascii=True)

        def _set() -> None:
            with sqlite3.connect(self._db_path) as conn:
                conn.execute(
                    """
                    UPDATE analyses
                    SET status = ?, updated_at = ?, result_json = ?, error = ''
                    WHERE id = ?
                    """,
                    ("SUCCEEDED", now, payload, id),
                )
                conn.commit()

        await asyncio.to_thread(_set)

    async def set_failed(self, *, id: str, error: str) -> None:
        now = _now_iso()

        def _set() -> None:
            with sqlite3.connect(self._db_path) as conn:
                conn.execute(
                    """
                    UPDATE analyses
                    SET status = ?, updated_at = ?, error = ?
                    WHERE id = ?
                    """,
                    ("FAILED", now, error, id),
                )
                conn.commit()

        await asyncio.to_thread(_set)

    async def get(self, *, id: str) -> AnalysisRecord | None:
        def _get() -> dict[str, Any] | None:
            with sqlite3.connect(self._db_path) as conn:
                cur = conn.execute(
                    """
                    SELECT id, git_url, ref, status, result_json, error
                    FROM analyses
                    WHERE id = ?
                    """,
                    (id,),
                )
                row = cur.fetchone()
                if not row:
                    return None
                return {
                    "id": row[0],
                    "git_url": row[1],
                    "ref": row[2],
                    "status": row[3],
                    "result_json": row[4],
                    "error": row[5] or "",
                }

        data = await asyncio.to_thread(_get)
        if data is None:
            return None

        result = None
        if data.get("result_json"):
            try:
                result = AnalysisResultData.from_json_dict(json.loads(data["result_json"]))
            except Exception:
                result = None

        return AnalysisRecord(
            id=str(data["id"]),
            git_url=str(data["git_url"]),
            ref=str(data["ref"]),
            status=str(data["status"]),
            result=result,
            error=str(data["error"]),
        )

File: services/analyzer-ts/src/store.ts
import Database from "better-sqlite3";
import { mkdirSync } from "node:fs";
import { dirname } from "node:path";
import type { AnalysisResultData } from "./models.js";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface AnalysisRecord {
  id: string;
  gitUrl: string;
  ref: string;
  status: string; // RUNNING | SUCCEEDED | FAILED
  result: AnalysisResultData | null;
  error: string;
}

export interface RepoSummaryRecord {
  gitUrl: string;
  ref: string;
  lastAnalysisId: string;
  lastStatus: string;
  lastUpdatedAt: string;
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function nowIso(): string {
  return new Date().toISOString();
}

// ---------------------------------------------------------------------------
// Store
// ---------------------------------------------------------------------------

export class SQLiteStore {
  private db: Database.Database | null = null;

  constructor(private dbPath: string) {}

  init(): void {
    mkdirSync(dirname(this.dbPath), { recursive: true });
    this.db = new Database(this.dbPath);
    this.db.pragma("journal_mode = WAL");
    this.db.exec(`
      CREATE TABLE IF NOT EXISTS analyses (
        id TEXT PRIMARY KEY,
        git_url TEXT NOT NULL,
        ref TEXT NOT NULL,
        status TEXT NOT NULL,
        created_at TEXT NOT NULL,
        updated_at TEXT NOT NULL,
        result_json TEXT,
        error TEXT
      )
    `);
    this.db.exec(`
      CREATE INDEX IF NOT EXISTS idx_analyses_repo_updated
      ON analyses (git_url, ref, updated_at)
    `);
    this.db.exec(`
      CREATE INDEX IF NOT EXISTS idx_analyses_updated
      ON analyses (updated_at)
    `);
  }

  private getDb(): Database.Database {
    if (!this.db) throw new Error("Store not initialized. Call init() first.");
    return this.db;
  }

  create(params: { id: string; gitUrl: string; ref: string }): void {
    const now = nowIso();
    this.getDb()
      .prepare(
        `INSERT INTO analyses (id, git_url, ref, status, created_at, updated_at, result_json, error)
         VALUES (?, ?, ?, ?, ?, ?, NULL, '')`,
      )
      .run(params.id, params.gitUrl, params.ref, "RUNNING", now, now);
  }

  setSucceeded(params: { id: string; result: AnalysisResultData }): void {
    const now = nowIso();
    const payload = JSON.stringify(params.result);
    this.getDb()
      .prepare(
        `UPDATE analyses
         SET status = ?, updated_at = ?, result_json = ?, error = ''
         WHERE id = ?`,
      )
      .run("SUCCEEDED", now, payload, params.id);
  }

  setFailed(params: { id: string; error: string }): void {
    const now = nowIso();
    this.getDb()
      .prepare(
        `UPDATE analyses
         SET status = ?, updated_at = ?, error = ?
         WHERE id = ?`,
      )
      .run("FAILED", now, params.error, params.id);
  }

  get(params: { id: string }): AnalysisRecord | null {
    const row = this.getDb()
      .prepare(
        `SELECT id, git_url, ref, status, result_json, error
         FROM analyses WHERE id = ?`,
      )
      .get(params.id) as
      | {
          id: string;
          git_url: string;
          ref: string;
          status: string;
          result_json: string | null;
          error: string | null;
        }
      | undefined;

    if (!row) return null;

    let result: AnalysisResultData | null = null;
    if (row.result_json) {
      try {
        result = JSON.parse(row.result_json) as AnalysisResultData;
      } catch {
        result = null;
      }
    }

    return {
      id: row.id,
      gitUrl: row.git_url,
      ref: row.ref,
      status: row.status,
      result,
      error: row.error ?? "",
    };
  }

  listRepos(params: {
    limit: number;
    offset: number;
  }): RepoSummaryRecord[] {
    const rows = this.getDb()
      .prepare(
        `SELECT git_url, ref, id, status, updated_at
         FROM (
           SELECT
             id, git_url, ref, status, updated_at, created_at,
             ROW_NUMBER() OVER (
               PARTITION BY git_url, ref
               ORDER BY updated_at DESC, created_at DESC, id DESC
             ) AS rn
           FROM analyses
         )
         WHERE rn = 1
         ORDER BY updated_at DESC, git_url ASC, ref ASC
         LIMIT ? OFFSET ?`,
      )
      .all(params.limit, params.offset) as Array<{
      git_url: string;
      ref: string;
      id: string;
      status: string;
      updated_at: string;
    }>;

    return rows.map((r) => ({
      gitUrl: r.git_url,
      ref: r.ref,
      lastAnalysisId: r.id,
      lastStatus: r.status,
      lastUpdatedAt: r.updated_at,
    }));
  }

  deleteRepo(params: { gitUrl: string; ref: string }): number {
    const info = this.getDb()
      .prepare(`DELETE FROM analyses WHERE git_url = ? AND ref = ?`)
      .run(params.gitUrl, params.ref);
    return info.changes;
  }
}
